### PR TITLE
Patch many smaller issues with the renderer

### PR DESCRIPTION
**Pull request by Alwinfy**
- ITextComponents are properly pipelined from book data to user space!
 - - Mixing of $() codes and vanilla style formatting is possible and handled gracefully! (Kinda)
 - Different overflow modes for text that goes off the page-- truncate, resize, overflow
 - Several cleanups! (if-else chains are gauche and so are ad-hoc string splitters)
 - API hooks to register Patchouli `$()` functions/commands! Reduced access at the moment but it's extensible

### CODE REVIEW

**Comment by williewillus** [comment]
Is there any way you can split this into smaller parts? It's really hard to review everything at once.

(by the way, one of the advantages of patches over github PR's is that you can review multiple diffs on the same branch at once, but alas, we're on github PR's)

**Comment by williewillus** [comment]
At least the command/function registration can be split off

If possible, every item in your bullet list should be a separate review unit.

**Review by williewillus — CHANGES_REQUESTED** [review]
^

File: src/main/resources/data/patchouli/patchouli_books/comprehensive_test_book/en_us/entries/deserialization/test.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    {
+      "type": "text",
+      "text": "Old-style $(l)formatting codes$(/l) should be supported, but the text on the following page should be localized!"
+    },
+    {
+      "type": "text",
+      "text": {
+        "translate": "patchouli.gui.lexicon.edition_str",
+        "with": [
+          "Really $(4)cool text that shouldn't leak its formatting!"
```
Comment (williewillus) [inline src/main/resources/data/patchouli/patchouli_books/comprehensive_test_book/en_us/entries/deserialization/test.json:16]: maybe clarify what the intended behaviour is, idk what is meant by "leak its formatting", leak to where?

File: src/main/java/vazkii/patchouli/client/book/text/SpanState.java [hunk 2]
```diff
@@ -83,4 +101,37 @@ public void reset() {
 	public Style peekStyle() {
 		return styleStack.getFirst();
 	}
+
+        // Represents the transformations on a single stack frame.
+        private class SpanPartialState {
+                // This is null iff this state represents the base state (i.e. nothing to merge)
+                private final Style mergeStyle;
```
Comment (williewillus) [inline src/main/java/vazkii/patchouli/client/book/text/SpanState.java]: use `@Nullable `

File: src/main/java/vazkii/patchouli/client/book/text/SpanState.java [hunk 3]
```diff
@@ -83,4 +101,37 @@ public void reset() {
...
+
+                public SpanPartialState(Style style) {
+                        this.mergeStyle = style;
+                }
+
+                public void addModification(Function<Style, Style> f) {
+                        if (transformations == null) {
+                                // We use a linked list because this list is likely to be very small
+                                // in order to skimp on space.
+                                transformations = new LinkedList<>();
```
Comment (williewillus) [inline src/main/java/vazkii/patchouli/client/book/text/SpanState.java]: does this matter that much? this is all thrown away after layout, right? same with the lazy init thing

File: src/main/java/vazkii/patchouli/client/book/text/SpanState.java [hunk 4]
```diff
@@ -21,6 +22,9 @@
 
 	private final Style baseStyle;
 	private final Deque<Style> styleStack = new ArrayDeque<>();
+        // This keeps a list of all modifications made to the style at each layer,
+        // so that when we replace the base style, we can reconstruct the style stack.
+	private final Deque<SpanPartialState> modifiersStack = new ArrayDeque<>();
```
Comment (williewillus) [inline src/main/java/vazkii/patchouli/client/book/text/SpanState.java:27]: consider pulling this + styleStack into a separate structure that ensures that they're consistent with each other. This seems fine right now but bugs might arise later if someone edits code touching styleStack without also knowing they have to do the same for modifiersStack.

File: src/main/resources/data/patchouli/patchouli_books/comprehensive_test_book/en_us/entries/deserialization/overflow_render.json [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    }, {
+      "type": "text",
+      "text": "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo. Nemo enim ipsam voluptatem quia voluptas sit aspernatur aut odit aut fugit, sed quia consequuntur magni dolores eos qui ratione voluptatem sequi nesciunt. Neque porro quisquam est, qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit, sed quia non numquam eius modi tempora incidunt ut labore et dolore magnam aliquam quaerat voluptatem. Ut enim ad minima veniam, quis nostrum exercitationem ullam corporis suscipit laboriosam, nisi ut aliquid ex ea commodi consequatur? Quis autem vel eum iure reprehenderit qui in ea voluptate velit esse quam nihil molestiae consequatur, vel illum qui dolorem eum fugiat quo voluptas nulla pariatur?"
```
Comment (williewillus) [inline src/main/resources/data/patchouli/patchouli_books/comprehensive_test_book/en_us/entries/deserialization/overflow_render.json:12]: ~~funny psi uncraftable copypasta time (or another of your choice)~~

**Review by williewillus — COMMENTED** [review]
looks fine overall

File: src/main/java/vazkii/patchouli/client/book/text/BookTextParser.java [hunk 6]
```diff
@@ -224,48 +224,36 @@ public String expandMacros(@Nullable String text) {
...
-				while (i < chars.length && chars[i] != ')') {
-					i++;
-				}
-
-				if (i == chars.length) {
-					spans.add(Span.error(state, "[ERROR: UNFINISHED COMMAND]"));
-					break;
-				}
+		while (matcher.find()) {
+			StringBuffer sb = new StringBuffer();
```
[the remark this answers is not in the document]
Comment (Alwinfy) [reply src/main/java/vazkii/patchouli/client/book/text/BookTextParser.java:253]: API limitation of Java  8, they add overloads for Matcher.appendReplacement(StringBuilder) in J9 but that's not our target

File: src/main/resources/data/patchouli/patchouli_books/comprehensive_test_book/en_us/entries/deserialization/test.json (hunk 1 above)
Comment (Alwinfy) [reply src/main/resources/data/patchouli/patchouli_books/comprehensive_test_book/en_us/entries/deserialization/test.json:16]: That is outdated! Let me fix that real quick

File: src/main/java/vazkii/patchouli/client/book/text/SpanState.java (hunk 3 above)
Comment (Alwinfy) [reply src/main/java/vazkii/patchouli/client/book/text/SpanState.java]: honestly, unless the user makes a _huge_ layout change, it shouldn't matter. i'm cheap anyway :wacko:

File: src/main/resources/data/patchouli/patchouli_books/comprehensive_test_book/en_us/entries/deserialization/overflow_render.json (hunk 5 above)
Comment (Alwinfy) [reply src/main/resources/data/patchouli/patchouli_books/comprehensive_test_book/en_us/entries/deserialization/overflow_render.json:12]: haha yes

File: src/main/java/vazkii/patchouli/common/base/PatchouliConfig.java [hunk 7]
```diff
@@ -64,6 +65,10 @@ public ClientLoader(ForgeConfigSpec.Builder builder) {
 					.comment("Set this to true to use Shift instead of Ctrl for the inventory quick lookup feature.")
 					.define("quickLookupShift", false);
 
+			overflowMode = builder
+					.comment("Set how text overflow should be coped with: overflow the text off the page, truncate overflowed text, or resize everything to fit.")
+					.defineEnum("textOverflowMode", TextOverflowMode.OVERFLOW);
```
Comment (Alwinfy) [inline src/main/java/vazkii/patchouli/common/base/PatchouliConfig.java:70]: Before merge, I'd like to know: Should this be set to OVERFLOW or RESIZE by default?

Comment (williewillus) [reply src/main/java/vazkii/patchouli/common/base/PatchouliConfig.java:70]: overflow just so it keeps existing behaviour. In a separate PR we could explore setting this on a book or entry specific basis.

Comment (williewillus) [reply src/main/java/vazkii/patchouli/common/base/PatchouliConfig.java:70]: and we can change the default later once people make use of it for a bit

**Pull request merged**